packages/snjs/lib/services/features_service.ts

```diff
@@ -0,0 +1,181 @@
+  }
+
+  public async loadWebSocketUrl(): Promise<void> {
+    const storedValue = await this.storageService.getValue(
+      StorageKey.WebSocketUrl
+    );
+    this.webSocketUrl =
+      storedValue ||
+      this.webSocketUrl ||
+      (window as {
+        _websocket_url?: string;
+      })._websocket_url;
+  }
+
+  public startWebSocketConnection(authToken: string): void {
+    if (this.webSocketUrl) {
+      this.webSocket = new WebSocket(
+        `${this.webSocketUrl}?authToken=Bearer+${authToken}`
+      );
+      this.webSocket.onmessage = this.onWebSocketMessage.bind(this);
+      this.webSocket.onclose = this.onWebSocketClose.bind(this);
+    }
+  }
+
+  public closeWebSocketConnection(): void {
+    this.webSocket?.close();
+  }
+
+  private async setRoles(roles: RoleName[]): Promise<void> {
+    this.roles = roles;
+    await this.storageService.setValue(StorageKey.UserRoles, this.roles);
+  }
+
+  private haveRolesChanged(roles: RoleName[]): boolean {
+    return (
+      roles.some((role) => !this.roles.includes(role)) ||
+      this.roles.some((role) => !roles.includes(role))
+    );
+  }
+
+  private async updateFeatures(userUuid: UuidString): Promise<void> {
+    const featuresResponse = await this.apiService.getUserFeatures(userUuid);
+    if (!featuresResponse.error && featuresResponse.data) {
+      await this.mapFeaturesToItems(
+        (featuresResponse as UserFeaturesResponse).data.features
+      );
+    }
+  }
+
+  private createItemDataForFeature(feature: Feature) {
+    return {
+      content_type: feature.contentType,
+      content: {
+        identifier: feature.identifier,
+        name: feature.name,
+        hosted_url: feature.url,
+        url: feature.url,
+        local_url: null,
+        area: feature.area,
+        package_info: feature,
+        valid_until: feature.expiresAt,
+      },
+      references: [],
+    };
+  }
+
+  private async mapFeaturesToItems(features: Feature[]): Promise<void> {
+    const currentItems = this.itemManager.getItems([
+      ContentType.Component,
+      ContentType.Theme,
+    ]);
+    const itemsToDeleteUuids = [];
+    const today = new Date();
+
+    for (const feature of features) {
+      const itemData = this.createItemDataForFeature(feature);
+      const existingItem = currentItems.find((item) => {
+        if (
+          item.content &&
+          typeof item.content !== 'string' &&
+          item.content.package_info
+        ) {
+          const itemIdentifier = item.content.package_info.identifier;
+          return itemIdentifier === feature.identifier && !item.deleted;
+        }
+        return false;
+      });
+      if (existingItem) {
+        if (feature.expiresAt! < today) {
+          itemsToDeleteUuids.push(existingItem.uuid);
+        } else {
+          await this.itemManager.changeComponent(existingItem.uuid, (mutator) => {
+            mutator.setContent(itemData);
+          });
+        }
+      } else if (feature.expiresAt! >= today) {
```

Comment:
I think this would be just an `else` since we create the item but it would be set in readonly mode if expired.

But this wouldn't apply to themes. Only components.

Reply:
> But this wouldn't apply to themes. Only components.

So components should follow the current logic and be deleted as opposed to set to read-only?

Reply:
Yes correct.